superduperdb/misc/runnable/runnable.py

```
@@ -0,0 +1,98 @@
+
+    def set(self):
+        super().set()
+        [c() for c in self.on_set]
+
+    def clear(self):
+        super().clear()
+        [c() for c in self.on_set]
+
+    def __bool__(self):
+        return self.is_set()
+
+
+class Runnable:
+    """A base class for things that start, run, finish, stop and join
+
+    An Runnable can also be used as a context manager:
+
+        with my_thread:
+            do_stuff()
+
+    means the same as
+
+        my_thread.start()
+        try:
+            do_stuff()
+            my_thread.finish()
+            my_thread.join()
+        finally:
+            my_thread.stop()
+
+    """
+
+    #: An Event that is set once this object is actually running
+    running: Event
+
+    #: An event that is set once this object is fully stopped
+    stopped: Event
+
+    def __init__(self):
+        self.running = Event()
+        self.stopped = Event()
+
+    def start(self):
+        """Start this object.
+
+        Note that self.running might not be immediately true after this method completes
+        """
+        self.running.set()
+
+    def stop(self):
+        """Stop as soon as possible. might not do anything, should never raise.
+
+        Note that self.stopped might not be immediately true after this method completes
+        """
+        self.running.clear()
+        self.stopped.set()
+
+    def finish(self):
+        """Request an orderly shutdown where all existing work is completed.
+
+        Note that self.stopped might not be immediately true after this method completes
+        """
+        self.stop()
+
+    def join(self, timeout: t.Optional[float] = None):
+        """Join this thread or process.  Might block indefinitely, might do nothing"""
+
+    def __enter__(self):
+        self.start()
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        try:
+            if exc_type is None:
+                self.finish()
+                self.join()
+        finally:
```

What's the rationale for discarding the exceptions here?

---

We pass exceptions right through!

In fact, there are two "exception" paths here, one invisible.

The `finally` path is always called, whether or not an `Exception` is thrown, and only prevents the exception from propagating if an exception is raised.  

But `stop` is specified to never raise an exception.  So `Exception`s are provably passed through that block.


--


But `__exit__` is also called at the end of the context when there's an exception.  In that case, `exc_type` is not `None` but contains the type of the exception.

By default, these exceptions are always propagated.

We could suppress such `Exception`s by returning "true" (say [the docs](https://docs.python.org/3/reference/datamodel.html#object.__exit__) though I bet you could get away with "truthy").